### DISCUSSION TITLE
Foreign key constraints are not allowed

### DISCUSSION THREAD

**Discussion by bigint** [planetscale/discussion#18] [FAQs] [answered]
I'm getting this error in planetscale but not in local DB

```
SQLSTATE[70100]: <<Unknown error>>: 1317 foreign key constraints are not allowed, see https://code.openark.org/blog/mysql/the-problem-with-mysql-foreign-key-constraints-in-online-schema-changes (SQL: alter table `tasks` add constraint `tasks_user_id_foreign` foreign key (`user_id`) references `users` (`id`) on delete cascade)
```

**Comment by lkburns** [comment]
Hey @bigint, our non-blocking schema change workflow can't support foreign key constraints so PlanetScale databases reject them. You can read more about the challenges of online schema changes with foreign key constraints here: https://planetscale.com/docs/learn/operating-without-foreign-key-constraints

**Reply by bigint** [reply]
So this platform won't support Laravel framework 🤔

**Reply by Gummibeer** [reply]
It does, but you can't use foreign keys in your migrations.

**Reply by bigint** [reply]
I'm working on a huge project that has 30+ tables, can we develop features without primary key @Gummibeer?

Here is the platform I'm working https://taskord.com

If so can we drop the existing FKs?

**Reply by Gummibeer** [reply]
Without PK? No one said that PKs aren't possible.
Foreign keys are nice but only a DB side mistake prevention. With proper observers and validation you get the same effect.

**Reply by NicolaGenesin** [reply]
Bumped on the same "issue". I would write this in bold somewhere. I wouldn't have considered PS if I knew this beforehand (I like the product, but I was actually porting a self hosted db and failed short because of this)

**Reply by pedrofcarvalho** [reply]
hey, I'm pretty new to working with DBs, especially located on a host server. 
I am working with PlanetScale and I was wondering exactly what I should use instead of FKs.

Thanks in advance!

**Reply by honekawa-yamazaki** [reply]
Hi, I'm a PlanetScale user. I'm very happy with my experience so far, so I'm gonna share my experience.
My application used MySQL at first, it has 70 tables with complicated relationships, a bunch of FKs, and the whole DB has over 10 million records. The largest table has 2M+ records.
When the DB was small (a few thousand to hundreds of thousands of records), there was no problem. But as the DB grows up, it gets slower when we perform a delete or update. The cascade mechanism blocks the DB for a long time, can say in minutes. The DB server management task became a nightmare. Of course, I have proper indexes and read replicas. But literally, the effort needed to maintain the database is really a pain in the a**.
At that point I decided that I need to give up the comfort of the foreign key constraint. I dropped all the foreign keys & write a constraint mechanism at the application level. For example, for cascade delete, I'll first mark the record as "deleted" and keep it stay in the DB but exclude it from all of the app queries. At the same time, I initiated a background job to start deleting related resources by chunk. Keep repeating that until the final "child" resource is deleted, and then perform an actual delete to the parent resource.
I know it's not an easy thing to do when your app is already working in production, and there is a lot of logic. But if you are really about to scale MySQL, it's an exchange that I think you must do. Let your app handles the cascade/relation asynchronously, and keep your database straight and simple as much as possible.
Then I migrated the DB to PlanetScale, and my life is like never been better. It's serverless, saving me 70% of the server cost (my estimation).
And, finally, I'd say if your DB isn't big enough or never gonna be (10M+ of records, for example), and you feel that it's a hassle to give up FKs. Then my advice is you can consider not doing it. Proper indexes and read replicas would handle the traffics very well. But if you love the joy of serverless and the scalability, then PlanetScale is the best, and in exchange, you will need to handle the cascade/relation on your own at the app level.

**Reply by EnkhAmar** [reply]
@honekawa-yamazaki Thanks for sharing your experience.

**Reply by MarcACard** [reply]
Hey @pedrofcarvalho, 

I found this blog - [Using MySQL with SQL Alchemy](https://planetscale.com/blog/using-mysql-with-sql-alchemy-hands-on-examples) by planetscale helpful in understanding how you can work with PlanetScale and relationships. The blog is focused on SQLAlchemy, a python orm, but its a demonstration of how you can manage a relationship in your code.

Also linked within the article is this page [Operating without foreign key constraints](https://planetscale.com/docs/learn/operating-without-foreign-key-constraints) from documentation

**Reply by danielo515** [reply]
I think it will be cool if you make a clear statement in the official docs that you don't need to remove the foreign key relations from your drizzle schemas to work with planetscale, because they will be rejected but it will still work

**Reply by mscoutermarsh** [reply]
Major update: https://planetscale.com/blog/announcing-foreign-key-constraints-support

**Comment by dogukan6durukan** [comment]
Hello, I want to drop a little note that if you guys are using Prisma with Planet-Scale which is very on demand and the way I love using it. You can maintain this Prisma schema below.

```prisma
datasource db {
  provider     = "mysql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
}

model User {
  id    Int    @id
  posts Post[]
}

model Post {
  id     Int  @id
  userId Int
  user   User @relation(fields: [userId], references: [id])
}
```
Set relation mode to `prisma` because there is a quote from the Prisma documentation :

>  The prisma relation mode is the default option for the MongoDB connector. It should also be set if you use a relational database that does not support foreign keys. For example, [**if you use PlanetScale**](https://www.prisma.io/docs/guides/database/using-prisma-with-planetscale#how-to-enable-emulation-of-referential-integrity) you should use the **prisma relation mode**.  

You can check out more from Prisma documentation. [DOCS](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/relation-mode#handle-relations-in-your-relational-database-with-the-foreignkeys-relation-mode)

**Reply by 205g0** [reply]
But isn't this just for Prisma? So it can detect a relationship between two entities and offer respective methods?! It does not enforce any constraints, means Primsa won't check if userId really exists on User. So IDK, if this helps in the context of our thread.

**Reply by dogukan6durukan** [reply]
It's not for Primsa. It's been a while on this thread but it's for both Planetscale and Prisma ORM. Some databases, such as MongoDB or Planetscale, do not support foreign keys and Prisma offers you prisma relation mode to emulate this. If you don't use Prisma mode you will also encounter this problem that I had and some people earlier...

**Comment by xcrap** [comment]
Oh, that's a shame. I've just registered to try the services and look really good and fast but that involves changing a lot of code in my projects. Wish there was another solution.

**Comment by dbussink** [comment, marked as answer]
We've announced support for foreign key constraints! https://planetscale.com/blog/announcing-foreign-key-constraints-support

**Reply by xcrap** [reply]
This is the way! :)
